advanced_topics.qmd

```diff
@@ -5,19 +5,217 @@
+We set the `startsWith()` argument `prefix` to "w", and that was passed to the Acero `starts_with()` option called `pattern`.
+As `startsWith()` is always case-sensitive, the Acero `starts_with()` option `ignore_case` has been set to `false`.
+
+### Bindings for aggregation functions
+
+The arrow R package also has bindings for aggregation functions which take multiple rows of data and calculate a single values.
+For example, whereas R has `sd()` for calculating the standard deviation of a vector, Acero has `stddev()`.
+
+| Feature                        | R (`sd`)                                                | Acero (`stddev`)                                                                        |
+|-------------------|----------------------|--------------------------------|
+| **Function Name**              | `sd`                                                    | `stddev`                                                                                |
+| **Function Options**           | `na.rm = FALSE`                                         | `ddof = 0`, `skip_nulls = FALSE`, `min_count = 1`                                       |
+| **Purpose**                    | Computes standard deviation                             | Computes standard deviation                                                             |
+| **Handling of Missing Values** | `na.rm`: Boolean, removes `NA` values (default `FALSE`) | `skip_nulls`: Boolean, removes `NULL` values (default `FALSE`)                          |
+| **Degrees of Freedom**         | Not available                                           | `ddof`: Integer, denominator degrees of freedom (default `0`)                           |
+| **Minimum Count**              | Not available                                           | `min_count`: Integer, returns `NULL` if fewer non-null values are present (default `1`) |
+
+There are also some differences in the options for these functions; while R's `sd` function has arguments `x`, the value to be passed in and `na.rm`, whether to remove `NA` values or not, Acero's stddev also has `ddof` - denominator degrees of freedom, `skip_nulls`, the equivalent of `na_rm`, and `min_count` - if there are fewer values than this, return `NULL`.
+
+```{r}
+#| label: sd_binding
+open_dataset("data/person/") |>
+  summarise(sd_age = sd(AGEP))
+```
+
+The output here doesn't show the function mapping as aggregation functions are handled differently from `select()`, `filter()`, and `mutate()`.
+The aggregation step is contained in the nested `.data` attribute of the query.
+This nesting ensures that the steps of the query are evaluated in this order.
+While `select()`, `filter()`, and `mutate()` can happen successively and be reflected in a single query step, any actions that happen on the data after `summarize()` (and, for that matter, joins) cannot be collapsed or rearranged like that.
```

Comment:
I added this paragraph, not sure it's totally clear. LMK if you want me to expand on it: could discuss in the "query optimizer" section above how the flattening of select/filter/mutate works, so then this would call back to that.

Also, should we introduce `collapse()`? If there were cases where that flattening were suboptimal, like if it led Acero to recompute a complex expression many times instead of doing it once and then mutating the result, it could be helpful. Though I haven't done the analysis to find a case where it would help.

Reply:
I made https://github.com/apache/arrow/issues/43894 to make this print better (if you looked at `.data`), seems easy enough, if we care.

Reply:
I guess we could be more explicit about exact why we can't just collapse or rearrange them - I think it's obvious to us but not necessarily to others.